distributed/scheduler.py

```diff
@@ -4242,6 +4242,10 @@ async def close(self, fast=False, close_workers=False):
         logger.info("Scheduler closing...")
         setproctitle("dask-scheduler [closing]")
 
+        await asyncio.gather(
+            *[plugin.before_close() for plugin in list(self.plugins.values())]
```

Comment:
Note that for the specific case we were looking at, we don't actually need this `before_close` hook: https://github.com/dask/distributed/issues/5660#issuecomment-1049275637

It does seem generally useful to me so I'm fine to see it added though. My one worry is about whether this will break existing plugins people have written. Do we require you to directly subclass `SchedulerPlugin`, or is it just an interface? Can we safely assume all plugin instances will have a `before_close` attribute? @fjetter @crusaderky thoughts?

I _might_ also say that a method called `before_close` should be the very first thing to run here (before `self.Status` gets changed or `self.finished` gets called).

Reply:
> It does seem generally useful to me so I'm fine to see it added though. My one worry is about whether this will break existing plugins people have written. Do we require you to directly subclass SchedulerPlugin, or is it just an interface? Can we safely assume all plugin instances will have a before_close attribute? @fjetter @crusaderky thoughts?

I've only seen subclasses of `SchedulerPlugin`. In this case adding the following shouldn't impact existing plugins (it doesn't on the local test suite). 

```python
class SchedulerPlugin:
  def before_close(self):
     pass
```

I would argue that SchedulerPlugin is actually more of a base class than an interface. If it was an interface, it would inherit from `ABC` and define all of it's methods as `@abstractmethod`, or just `raise NotImplementedError` in all of them. For instance, the [WorkStealing](https://github.com/dask/distributed/blob/2fbf9ebef4ea862b72bd221f63d2785016d0086c/distributed/stealing.py#L52) plugin doesn't implement `update_graph`, but merely invokes the base method which does nothing.

Reply:
> I might also say that a method called before_close should be the very first thing to run here (before self.Status gets changed or self.finished gets called).

The `before_close` calls are now called in the following order:

```python
    async def close(self, fast=False, close_workers=False):
        parent: SchedulerState = cast(SchedulerState, self)
        if self.status in (Status.closing, Status.closed):
            await self.finished()
            return

        await asyncio.gather(
            *[plugin.before_close() for plugin in list(self.plugins.values())]
        )

        self.status = Status.closing
```

I think if `before_close` was called before `self.finished()` it could be invoked multiple times which is not a desirable scenario.

Reply:
> Can we safely assume all plugin instances will have a before_close attribute? @fjetter @crusaderky thoughts?

All prior art is blindly invoking plugin methods without testing if they're actually there. 

XREF https://github.com/dask/distributed/pull/6008